pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 			Name:      "distributor_forward_samples_total",
 			Help:      "The total number of samples the Distributor forwarded.",
 		}),
+		exemplarsTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Namespace: "cortex",
+			Name:      "distributor_forward_exemplars_total",
+			Help:      "The total number of exemplars the Distributor forwarded.",
+		}),
 		requestLatencyHistogram: promauto.With(reg).NewHistogram(prometheus.HistogramOpts{
 			Namespace: "cortex",
 			Name:      "distributor_forward_requests_latency_seconds",
 			Help:      "The client-side latency of requests to forward metrics made by the Distributor.",
 			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
 		}),
 	}
+
+	f.workerWg.Add(f.cfg.RequestConcurrency)
+	for i := 0; i < f.cfg.RequestConcurrency; i++ {
+		go f.worker()
+	}
+
+	return f
 }
 
-func (r *forwarder) NewRequest(ctx context.Context, tenant string, rules validation.ForwardingRules) Request {
-	return &request{
-		ctx:    ctx,
-		client: &r.client, // http client should be re-used so open connections get re-used.
-		pools:  &r.pools,
+func (f *forwarder) Stop() {
+	close(f.reqCh)
+	f.workerWg.Wait()
+}
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+func (f *forwarder) Forward(ctx context.Context, rules validation.ForwardingRules, in []mimirpb.PreallocTimeseries) (TimeseriesCounts, []mimirpb.PreallocTimeseries, chan error) {
```

Comment:
I would like to see the behavior of this function documented, in order to understand the code and the usage better.

Currently it looks like it will enqueue requests that eventually will return an error, if any, through the returned chan. 

This allows the caller to do other operations, i.e. `PushWithCleanup`, while this is forwarding.

However, I also can see that this _will_ block if too many requests are already enqueued, so its behavior isn't inconsistent and I don't know what we should expect.

Reply:
Makes sense, how is this:

https://github.com/grafana/mimir/blob/56f1d28a6f28782e01f135630bbe3dff64c20b83/pkg/distributor/forwarding/forwarding.go#L127-L140

Reply:
Thank you!